File: vibe-trading-saas/worker/src/hm_worker/providers.py
"""LLM provider catalog — mirrors Tradi/agent/src/providers/llm_providers.json.

The worker must not import the engine (D1), so this module provides a
self-contained copy of the provider catalog.  The Docker image also installs
the engine, so we attempt to read the canonical JSON first; if that fails
(local dev, test, path change), we fall back to the embedded constant.

Each entry tells ``_build_env()`` which environment variables to set so the
engine's ``_sync_provider_env()`` picks the right LLM adapter at runtime.
"""

from __future__ import annotations

import json
import logging
from dataclasses import dataclass
from pathlib import Path
from typing import Sequence

log = logging.getLogger(__name__)


@dataclass(frozen=True)
class ProviderEntry:
    """One supported LLM provider."""

    name: str
    label: str
    api_key_env: str | None
    base_url_env: str | None
    default_model: str
    default_base_url: str
    api_key_required: bool

    @property
    def is_base_url_provider(self) -> bool:
        """True for providers that store a URL instead of an API key (ollama, copilot)."""
        return not self.api_key_required


# Paths to try when loading the canonical JSON, ordered by likelihood.
_CATALOG_JSON_PATHS = [
    Path("/app/Tradi/agent/src/providers/llm_providers.json"),  # Docker image
]

# Embedded fallback — synced from llm_providers.json 2026-08-22.
# openai-codex is excluded (requires OAuth, not viable for SaaS BYOK).
_EMBEDDED_CATALOG: list[dict] = [
    {"name": "openrouter", "label": "OpenRouter", "api_key_env": "OPENROUTER_API_KEY", "base_url_env": "OPENROUTER_BASE_URL", "default_model": "deepseek/deepseek-v4-pro", "default_base_url": "https://openrouter.ai/api/v1", "api_key_required": True},
    {"name": "requesty", "label": "Requesty", "api_key_env": "REQUESTY_API_KEY", "base_url_env": "REQUESTY_BASE_URL", "default_model": "openai/gpt-4o-mini", "default_base_url": "https://router.requesty.ai/v1", "api_key_required": True},
    {"name": "openai", "label": "OpenAI", "api_key_env": "OPENAI_API_KEY", "base_url_env": "OPENAI_BASE_URL", "default_model": "gpt-5.5", "default_base_url": "https://api.openai.com/v1", "api_key_required": True},
    {"name": "anthropic", "label": "Anthropic", "api_key_env": "ANTHROPIC_API_KEY", "base_url_env": "ANTHROPIC_BASE_URL", "default_model": "claude-sonnet-4-6", "default_base_url": "https://api.anthropic.com", "api_key_required": True},
    {"name": "deepseek", "label": "DeepSeek", "api_key_env": "DEEPSEEK_API_KEY", "base_url_env": "DEEPSEEK_BASE_URL", "default_model": "deepseek-v4-pro", "default_base_url": "https://api.deepseek.com/v1", "api_key_required": True},
    {"name": "siliconflow-cn", "label": "SiliconFlow (CN)", "api_key_env": "SILICONFLOW_API_KEY", "base_url_env": "SILICONFLOW_BASE_URL", "default_model": "deepseek-ai/DeepSeek-V3.1-Terminus", "default_base_url": "https://api.siliconflow.cn/v1", "api_key_required": True},
    {"name": "siliconflow-global", "label": "SiliconFlow (Global)", "api_key_env": "SILICONFLOW_GLOBAL_API_KEY", "base_url_env": "SILICONFLOW_GLOBAL_BASE_URL", "default_model": "deepseek-ai/DeepSeek-V3.1-Terminus", "default_base_url": "https://api.siliconflow.com/v1", "api_key_required": True},
    {"name": "nvidia", "label": "NVIDIA NIM", "api_key_env": "NVIDIA_API_KEY", "base_url_env": "NVIDIA_BASE_URL", "default_model": "nvidia/nemotron-3-ultra-550b-a55b", "default_base_url": "https://integrate.api.nvidia.com/v1", "api_key_required": True},
    {"name": "gemini", "label": "Gemini", "api_key_env": "GEMINI_API_KEY", "base_url_env": "GEMINI_BASE_URL", "default_model": "gemini-3.5-flash", "default_base_url": "https://generativelanguage.googleapis.com/v1beta/openai/", "api_key_required": True},
    {"name": "groq", "label": "Groq", "api_key_env": "GROQ_API_KEY", "base_url_env": "GROQ_BASE_URL", "default_model": "meta-llama/llama-4-maverick-17b-128e-instruct", "default_base_url": "https://api.groq.com/openai/v1", "api_key_required": True},
    {"name": "novita", "label": "Novita AI", "api_key_env": "NOVITA_API_KEY", "base_url_env": "NOVITA_BASE_URL", "default_model": "moonshotai/kimi-k3", "default_base_url": "https://api.novita.ai/openai", "api_key_required": True},
    {"name": "dashscope", "label": "DashScope / Qwen", "api_key_env": "DASHSCOPE_API_KEY", "base_url_env": "DASHSCOPE_BASE_URL", "default_model": "qwen-plus-latest", "default_base_url": "https://dashscope.aliyuncs.com/compatible-mode/v1", "api_key_required": True},
    {"name": "qwen", "label": "Qwen", "api_key_env": "DASHSCOPE_API_KEY", "base_url_env": "DASHSCOPE_BASE_URL", "default_model": "qwen-plus-latest", "default_base_url": "https://dashscope.aliyuncs.com/compatible-mode/v1", "api_key_required": True},
    {"name": "zhipu", "label": "Zhipu", "api_key_env": "ZHIPU_API_KEY", "base_url_env": "ZHIPU_BASE_URL", "default_model": "glm-5.1", "default_base_url": "https://open.bigmodel.cn/api/paas/v4", "api_key_required": True},
    {"name": "glm", "label": "GLM (Zhipu)", "api_key_env": "ZHIPU_API_KEY", "base_url_env": "ZHIPU_BASE_URL", "default_model": "glm-5.1", "default_base_url": "https://open.bigmodel.cn/api/paas/v4", "api_key_required": True},
    {"name": "moonshot", "label": "Moonshot / Kimi", "api_key_env": "MOONSHOT_API_KEY", "base_url_env": "MOONSHOT_BASE_URL", "default_model": "kimi-k2.6", "default_base_url": "https://api.moonshot.ai/v1", "api_key_required": True},
    {"name": "kimi-coding", "label": "Kimi for Coding", "api_key_env": "KIMI_CODING_API_KEY", "base_url_env": "KIMI_CODING_BASE_URL", "default_model": "kimi-for-coding", "default_base_url": "https://api.kimi.com/coding/v1", "api_key_required": True},
    {"name": "minimax", "label": "MiniMax", "api_key_env": "MINIMAX_API_KEY", "base_url_env": "MINIMAX_BASE_URL", "default_model": "MiniMax-M3", "default_base_url": "https://api.minimax.io/v1", "api_key_required": True},
    {"name": "mimo", "label": "Xiaomi MIMO", "api_key_env": "MIMO_API_KEY", "base_url_env": "MIMO_BASE_URL", "default_model": "MiMo-72B-A27B", "default_base_url": "https://api.xiaomimimo.com/v1", "api_key_required": True},
    {"name": "spark", "label": "iFlytek Spark", "api_key_env": "SPARK_API_KEY", "base_url_env": "SPARK_BASE_URL", "default_model": "4.0Ultra", "default_base_url": "https://spark-api-open.xf-yun.com/v1", "api_key_required": True},
    {"name": "zai", "label": "Z.ai", "api_key_env": "ZAI_API_KEY", "base_url_env": "ZAI_BASE_URL", "default_model": "glm-5.1", "default_base_url": "https://api.z.ai/api/coding/paas/v4", "api_key_required": True},
    {"name": "modelscope", "label": "ModelScope", "api_key_env": "MODELSCOPE_API_KEY", "base_url_env": "MODELSCOPE_BASE_URL", "default_model": "Qwen/Qwen3.5-27B", "default_base_url": "https://api-inference.modelscope.cn/v1", "api_key_required": True},
    {"name": "ollama", "label": "Ollama", "api_key_env": None, "base_url_env": "OLLAMA_BASE_URL", "default_model": "qwen2.5:32b", "default_base_url": "http://localhost:11434", "api_key_required": False},
    {"name": "copilot", "label": "GitHub Copilot SDK", "api_key_env": "COPILOT_GITHUB_TOKEN", "base_url_env": "COPILOT_BASE_URL", "default_model": "claude-sonnet-5", "default_base_url": "https://api.githubcopilot.com", "api_key_required": False},
]

# Resolution priority: deepseek first (backward compat with existing users),
# then popular API-key providers, regional providers, base-URL providers last.
_RESOLUTION_ORDER = [
    "deepseek",
    "openai",
    "anthropic",
    "gemini",
    "groq",
    "openrouter",
    "requesty",
    "nvidia",
    "novita",
    "siliconflow-cn",
    "siliconflow-global",
    "dashscope",
    "qwen",
    "zhipu",
    "glm",
    "moonshot",
    "kimi-coding",
    "minimax",
    "mimo",
    "spark",
    "zai",
    "modelscope",
    "ollama",
    "copilot",
]


def _load_catalog() -> tuple[dict[str, ProviderEntry], list[str]]:
    """Load the provider catalog, returning ``(lookup, resolution_order)``."""
    raw: list[dict] = _EMBEDDED_CATALOG

    for path in _CATALOG_JSON_PATHS:
        try:
            raw = json.loads(path.read_text())
            log.debug("loaded provider catalog from %s (%d entries)", path, len(raw))
            break
        except (OSError, json.JSONDecodeError):
            continue

    lookup: dict[str, ProviderEntry] = {}
    for entry in raw:
        name = entry.get("name", "")
        if not name:
            continue
        # Skip openai-codex — requires OAuth, not viable for SaaS BYOK.
        if name == "openai-codex":
            continue
        lookup[name] = ProviderEntry(
            name=name,
            label=entry.get("label", name),
            api_key_env=entry.get("api_key_env"),
            base_url_env=entry.get("base_url_env"),
            default_model=entry.get("default_model", ""),
            default_base_url=entry.get("default_base_url", ""),
            api_key_required=entry.get("api_key_required", True),
        )

    # Build resolution order: prefer the defined priority list, then append
    # any new providers from the catalog that weren't in it.
    order = [n for n in _RESOLUTION_ORDER if n in lookup]
    for name in lookup:
        if name not in order:
            order.append(name)

    return lookup, order


PROVIDER_CATALOG, RESOLUTION_ORDER = _load_catalog()


def get_provider(name: str) -> ProviderEntry | None:
    """Look up a provider by name.  Returns ``None`` if unknown."""
    return PROVIDER_CATALOG.get(name)


def all_provider_names() -> Sequence[str]:
    """All supported provider names in resolution order."""
    return RESOLUTION_ORDER

File: Tradi/frontend/src/lib/providers.ts
/**
 * LLM provider catalog — frontend mirror of llm_providers.json.
 *
 * Defines every LLM provider the platform supports for BYOK (bring your own
 * key). The worker uses the same catalog (providers.py) to resolve the user's
 * configured credential at run time.
 *
 * openai-codex is excluded — it requires OAuth, which isn't viable for our
 * SaaS BYOK model where the worker runs the credential server-side.
 */

export interface LlmProvider {
  /** Machine name — matches the DB provider column and llm_providers.json "name". */
  name: string;
  /** Human-readable display label. */
  label: string;
  /** True if this provider uses an API key; false if it uses a base URL. */
  apiKeyRequired: boolean;
  /** Placeholder for the input field. */
  placeholder: string;
  /** Input type: "password" for API keys, "url" for base URLs. */
  inputType: "password" | "url";
  /** Optional help URL where users can get their key. */
  helpUrl?: string;
  /** Short description shown under the provider name. */
  description: string;
  /** Whether this is a "featured" / popular provider (shown first in the UI). */
  featured: boolean;
  /** Category for grouping in the UI. */
  category: "popular" | "router" | "regional" | "self-hosted";
}

/**
 * All supported providers, ordered for display.
 * Popular providers first, then routers, then regional, then self-hosted.
 */
export const LLM_PROVIDERS: LlmProvider[] = [
  // ─── Popular ──────────────────────────────────────────────
  {
    name: "deepseek",
    label: "DeepSeek",
    apiKeyRequired: true,
    placeholder: "sk-...",
    inputType: "password",
    helpUrl: "https://platform.deepseek.com/api_keys",
    description: "DeepSeek AI — recommended for trading research.",
    featured: true,
    category: "popular",
  },
  {
    name: "openai",
    label: "OpenAI",
    apiKeyRequired: true,
    placeholder: "sk-...",
    inputType: "password",
    helpUrl: "https://platform.openai.com/api-keys",
    description: "GPT-5.5 and other OpenAI models.",
    featured: true,
    category: "popular",
  },
  {
    name: "anthropic",
    label: "Anthropic",
    apiKeyRequired: true,
    placeholder: "sk-ant-...",
    inputType: "password",
    helpUrl: "https://console.anthropic.com/settings/keys",
    description: "Claude models via the Anthropic API.",
    featured: true,
    category: "popular",
  },
  {
    name: "gemini",
    label: "Gemini",
    apiKeyRequired: true,
    placeholder: "AIza...",
    inputType: "password",
    helpUrl: "https://aistudio.google.com/apikey",
    description: "Google Gemini models.",
    featured: true,
    category: "popular",
  },
  {
    name: "groq",
    label: "Groq",
    apiKeyRequired: true,
    placeholder: "gsk_...",
    inputType: "password",
    helpUrl: "https://console.groq.com/keys",
    description: "Ultra-fast inference — Llama, Mixtral, and more.",
    featured: true,
    category: "popular",
  },

  // ─── Routers ──────────────────────────────────────────────
  {
    name: "openrouter",
    label: "OpenRouter",
    apiKeyRequired: true,
    placeholder: "sk-or-...",
    inputType: "password",
    helpUrl: "https://openrouter.ai/keys",
    description: "Access 200+ models through one API key.",
    featured: true,
    category: "router",
  },
  {
    name: "requesty",
    label: "Requesty",
    apiKeyRequired: true,
    placeholder: "Your Requesty API key",
    inputType: "password",
    helpUrl: "https://requesty.ai",
    description: "AI model router with smart routing.",
    featured: false,
    category: "router",
  },

  // ─── Regional / Specialized ───────────────────────────────
  {
    name: "nvidia",
    label: "NVIDIA NIM",
    apiKeyRequired: true,
    placeholder: "nvapi-...",
    inputType: "password",
    helpUrl: "https://build.nvidia.com",
    description: "NVIDIA NIM inference microservices.",
    featured: false,
    category: "regional",
  },
  {
    name: "novita",
    label: "Novita AI",
    apiKeyRequired: true,
    placeholder: "Your Novita API key",
    inputType: "password",
    helpUrl: "https://novita.ai",
    description: "Affordable model inference.",
    featured: false,
    category: "regional",
  },
  {
    name: "siliconflow-cn",
    label: "SiliconFlow (CN)",
    apiKeyRequired: true,
    placeholder: "Your SiliconFlow API key",
    inputType: "password",
    helpUrl: "https://siliconflow.cn",
    description: "SiliconFlow China region.",
    featured: false,
    category: "regional",
  },
  {
    name: "siliconflow-global",
    label: "SiliconFlow (Global)",
    apiKeyRequired: true,
    placeholder: "Your SiliconFlow API key",
    inputType: "password",
    helpUrl: "https://siliconflow.com",
    description: "SiliconFlow global region.",
    featured: false,
    category: "regional",
  },
  {
    name: "dashscope",
    label: "DashScope / Qwen",
    apiKeyRequired: true,
    placeholder: "Your DashScope API key",
    inputType: "password",
    helpUrl: "https://dashscope.console.aliyun.com",
    description: "Alibaba Cloud DashScope — Qwen models.",
    featured: false,
    category: "regional",
  },
  {
    name: "qwen",
    label: "Qwen",
    apiKeyRequired: true,
    placeholder: "Your DashScope API key",
    inputType: "password",
    helpUrl: "https://dashscope.console.aliyun.com",
    description: "Qwen models via DashScope.",
    featured: false,
    category: "regional",
  },
  {
    name: "zhipu",
    label: "Zhipu",
    apiKeyRequired: true,
    placeholder: "Your Zhipu API key",
    inputType: "password",
    helpUrl: "https://open.bigmodel.cn",
    description: "Zhipu AI — GLM models.",
    featured: false,
    category: "regional",
  },
  {
    name: "glm",
    label: "GLM (Zhipu)",
    apiKeyRequired: true,
    placeholder: "Your Zhipu API key",
    inputType: "password",
    helpUrl: "https://open.bigmodel.cn",
    description: "GLM models via Zhipu.",
    featured: false,
    category: "regional",
  },
  {
    name: "moonshot",
    label: "Moonshot / Kimi",
    apiKeyRequired: true,
    placeholder: "Your Moonshot API key",
    inputType: "password",
    helpUrl: "https://platform.moonshot.cn",
    description: "Moonshot AI — Kimi models.",
    featured: false,
    category: "regional",
  },
  {
    name: "kimi-coding",
    label: "Kimi for Coding",
    apiKeyRequired: true,
    placeholder: "Your Kimi Coding API key",
    inputType: "password",
    helpUrl: "https://platform.moonshot.cn",
    description: "Kimi optimized for code generation.",
    featured: false,
    category: "regional",
  },
  {
    name: "minimax",
    label: "MiniMax",
    apiKeyRequired: true,
    placeholder: "Your MiniMax API key",
    inputType: "password",
    helpUrl: "https://platform.minimax.io",
    description: "MiniMax AI models.",
    featured: false,
    category: "regional",
  },
  {
    name: "mimo",
    label: "Xiaomi MIMO",
    apiKeyRequired: true,
    placeholder: "Your MIMO API key",
    inputType: "password",
    helpUrl: "https://xiaomimimo.com",
    description: "Xiaomi MIMO large language models.",
    featured: false,
    category: "regional",
  },
  {
    name: "spark",
    label: "iFlytek Spark",
    apiKeyRequired: true,
    placeholder: "Your Spark API key",
    inputType: "password",
    helpUrl: "https://xinghuo.xfyun.cn",
    description: "iFlytek Spark AI models.",
    featured: false,
    category: "regional",
  },
  {
    name: "zai",
    label: "Z.ai",
    apiKeyRequired: true,
    placeholder: "Your Z.ai API key",
    inputType: "password",
    helpUrl: "https://z.ai",
    description: "Z.ai model API.",
    featured: false,
    category: "regional",
  },
  {
    name: "modelscope",
    label: "ModelScope",
    apiKeyRequired: true,
    placeholder: "Your ModelScope API key",
    inputType: "password",
    helpUrl: "https://modelscope.cn",
    description: "Alibaba ModelScope inference.",
    featured: false,
    category: "regional",
  },

  // ─── Self-hosted ──────────────────────────────────────────
  {
    name: "ollama",
    label: "Ollama",
    apiKeyRequired: false,
    placeholder: "http://localhost:11434",
    inputType: "url",
    helpUrl: "https://ollama.com",
    description: "Self-hosted — no API key needed, just a URL.",
    featured: true,
    category: "self-hosted",
  },
  {
    name: "copilot",
    label: "GitHub Copilot",
    apiKeyRequired: false,
    placeholder: "https://api.githubcopilot.com",
    inputType: "url",
    helpUrl: "https://github.com/features/copilot",
    description: "GitHub Copilot SDK — base URL.",
    featured: false,
    category: "self-hosted",
  },
];

/** Lookup map for O(1) access by name. */
export const PROVIDER_MAP = new Map(LLM_PROVIDERS.map((p) => [p.name, p]));

/** Featured providers (shown in the main grid). */
export const FEATURED_PROVIDERS = LLM_PROVIDERS.filter((p) => p.featured);

/** Non-featured providers (shown in the expandable "More providers" section). */
export const MORE_PROVIDERS = LLM_PROVIDERS.filter((p) => !p.featured);

/** All valid provider names. */
export const PROVIDER_NAMES = LLM_PROVIDERS.map((p) => p.name);
